File: skills/agent-harness/scripts/prepare_action_request.py
#!/usr/bin/env python3
"""Materialize one exact private action request from an approved grant and lease."""
from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import stat
from typing import Any

import check_authorization


class RequestError(ValueError):
    pass


def _load(path: Path, label: str, run_root: Path) -> Any:
    if (
        not path.is_absolute()
        or path.is_symlink()
        or not path.is_file()
        or path.parent.resolve(strict=True) != run_root
    ):
        raise RequestError(f"{label} must be a regular file directly under the run root")
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as error:
        raise RequestError(f"{label} is not readable JSON") from error


def prepare(
    *,
    authorization_path: Path,
    receipt_path: Path,
    descriptor_path: Path,
    health_report_path: Path,
    output_path: Path,
    run_root: Path,
    grant_id: str,
    target: str,
    paths: list[str],
    spec_checkpoint_path: Path | None = None,
    apple_action_path: Path | None = None,
    apple_observation_path: Path | None = None,
) -> dict[str, Any]:
    canonical_root = run_root.resolve(strict=True)
    if run_root.is_symlink() or not canonical_root.is_dir():
        raise RequestError("run root must be an existing non-symlink directory")
    if (
        not output_path.is_absolute()
        or output_path.parent.resolve(strict=True) != canonical_root
        or output_path.exists()
        or output_path.is_symlink()
    ):
        raise RequestError("output must be a new regular file directly under the run root")
    authorization = _load(authorization_path, "authorization", canonical_root)
    receipt = _load(receipt_path, "coordinator receipt", canonical_root)
    descriptor = _load(descriptor_path, "resource descriptor", canonical_root)
    health_report = _load(health_report_path, "health report", canonical_root)
    if not isinstance(authorization, dict):
        raise RequestError("authorization must contain an object")
    errors = check_authorization.validate_authorization(authorization)
    if errors:
        raise RequestError("authorization is invalid: " + "; ".join(sorted(set(errors))))
    grants = [
        grant for grant in authorization.get("action_grants", [])
        if isinstance(grant, dict) and grant.get("grant_id") == grant_id
    ]
    if len(grants) != 1:
        raise RequestError("grant_id must resolve exactly once")
    grant = grants[0]
    if grant.get("target") is not None and target != grant.get("target"):
        raise RequestError("target differs from the exact grant target")
    if not isinstance(receipt, dict) or set(receipt) != check_authorization.COORDINATOR_RECEIPT_FIELDS:
        raise RequestError("coordinator receipt shape is invalid")
    if receipt.get("owner_run_id") != authorization.get("run_id"):
        raise RequestError("coordinator receipt belongs to another run")
    if receipt.get("owner_actor") != authorization.get("selected_writer"):
        raise RequestError("coordinator receipt belongs to another writer")
    if receipt.get("resource_key") != grant.get("resource_key"):
        raise RequestError("coordinator receipt resource differs from the grant")
    if not isinstance(descriptor, dict) or not descriptor:
        raise RequestError("resource descriptor must be a non-empty object")
    try:
        descriptor = check_authorization.resource_coordinator.normalize_descriptor(
            str(receipt.get("resource")), descriptor
        )
        descriptor_sha256 = (
            check_authorization.resource_coordinator.descriptor_sha256(
                str(receipt.get("resource")), descriptor
            )
        )
    except check_authorization.resource_coordinator.CoordinatorError as error:
        raise RequestError(f"resource descriptor is invalid: {error.code}") from error
    if receipt.get("descriptor_sha256") != descriptor_sha256:
        raise RequestError("resource descriptor digest differs from the receipt")
    if not paths or len(paths) != len(set(paths)) or any(
        not isinstance(path, str) or not path for path in paths
    ):
        raise RequestError("paths must be non-empty unique strings")
    spec_checkpoint_sha256 = None
    if spec_checkpoint_path is not None:
        checkpoint = _load(spec_checkpoint_path, "Spec Kit checkpoint", canonical_root)
        spec_checkpoint_sha256 = check_authorization.canonical_sha256(checkpoint)
    apple_observation_sha256 = None
    if apple_observation_path is not None:
        observation = _load(apple_observation_path, "Apple observation", canonical_root)
        apple_observation_sha256 = check_authorization.canonical_sha256(observation)
    apple_action = None
    if apple_action_path is not None:
        apple_action = _load(apple_action_path, "Apple action", canonical_root)
    if str(grant.get("action", "")).startswith("apple.") and not isinstance(apple_action, dict):
        raise RequestError("Apple grants require --apple-action")
    if not str(grant.get("action", "")).startswith("apple.") and apple_action is not None:
        raise RequestError("non-Apple grants cannot include --apple-action")
    request = {
        "run_id": authorization["run_id"],
        "authorization_id": authorization["authorization_id"],
        "authorization_hash": check_authorization.authorization_hash(authorization),
        "delivery_target": authorization["delivery_target"],
        "system": grant["system"],
        "action": grant["action"],
        "target": target,
        "grant_id": grant["grant_id"],
        "idempotency_key": grant["idempotency_key"],
        "repository": authorization["repository"],
        "spec_snapshot_sha256": (
            authorization.get("spec_kit", {}).get("snapshot_sha256")
            if isinstance(authorization.get("spec_kit"), dict)
            else None
        ),
        "paths": paths,
        "apple": apple_action,
        "lease_id": receipt["lease_id"],
        "lease_owner": receipt["owner_actor"],
        "lease_resource": receipt["resource"],
        "lease_resource_key": receipt["resource_key"],
        "resource_descriptor": descriptor,
        "coordinator_receipt": receipt,
        "operation": grant["operation"],
        "operation_input": grant["operation_input"],
        "constraint_sha256": grant["constraint_sha256"],
        "phase": grant["phase"],
        "spec_checkpoint_sha256": spec_checkpoint_sha256,
        "apple_observation_sha256": apple_observation_sha256,
        "writer_actor": authorization["selected_writer"],
        "health_report_sha256": "sha256:"
        + check_authorization.canonical_sha256(health_report),
    }
    missing = check_authorization.REQUEST_FIELDS - set(request)
    extra = set(request) - check_authorization.REQUEST_FIELDS
    if missing or extra:
        raise RequestError("generated request field set drifted from the installed contract")
    descriptor_fd = os.open(
        output_path,
        os.O_WRONLY | os.O_CREAT | os.O_EXCL,
        stat.S_IRUSR | stat.S_IWUSR,
    )
    try:
        with os.fdopen(descriptor_fd, "w", encoding="utf-8") as handle:
            json.dump(request, handle, indent=2, sort_keys=True)
            handle.write("\n")
            handle.flush()
            os.fsync(handle.fileno())
    except Exception:
        try:
            output_path.unlink()
        except OSError:
            pass
        raise
    return request


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--authorization", type=Path, required=True)
    parser.add_argument("--receipt", type=Path, required=True)
    parser.add_argument("--resource-descriptor", type=Path, required=True)
    parser.add_argument("--health-report", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    parser.add_argument("--run-root", type=Path, required=True)
    parser.add_argument("--grant-id", required=True)
    parser.add_argument("--target", required=True)
    parser.add_argument("--path", action="append", dest="paths", required=True)
    parser.add_argument("--spec-checkpoint", type=Path)
    parser.add_argument("--apple-action", type=Path)
    parser.add_argument("--apple-observation", type=Path)
    arguments = parser.parse_args()
    try:
        request = prepare(
            authorization_path=arguments.authorization,
            receipt_path=arguments.receipt,
            descriptor_path=arguments.resource_descriptor,
            health_report_path=arguments.health_report,
            output_path=arguments.output,
            run_root=arguments.run_root,
            grant_id=arguments.grant_id,
            target=arguments.target,
            paths=arguments.paths,
            spec_checkpoint_path=arguments.spec_checkpoint,
            apple_action_path=arguments.apple_action,
            apple_observation_path=arguments.apple_observation,
        )
    except (RequestError, OSError) as error:
        print(json.dumps({"status": "blocked", "reason": str(error)}, sort_keys=True))
        return 2
    print(json.dumps({
        "status": "ok",
        "output": str(arguments.output.resolve(strict=True)),
        "grant_id": request["grant_id"],
        "authorization_hash": request["authorization_hash"],
    }, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: skills/agent-harness/verification/Sources/AppleVerificationCore/PrepareActionRequest.swift
import Darwin
import Foundation

public enum PrepareActionRequest {
  public static func run(arguments: [String], context: RuntimeContext) throws -> Int32 {
    try PrepareActionRequestCommand.run(arguments: arguments, context: context)
  }
  public static func prepare(
    authorizationPath: URL, receiptPath: URL, descriptorPath: URL,
    healthReportPath: URL, outputPath: URL, runRoot: URL,
    grantID: String, target: String, paths: [String],
    specCheckpointPath: URL? = nil, appleActionPath: URL? = nil,
    appleObservationPath: URL? = nil, context: RuntimeContext
  ) throws -> [String: Any] {
    let root = runRoot.resolvingSymlinksInPath().standardizedFileURL
    guard !isSymlink(runRoot), isDirectory(root) else {
      throw VerificationError.invalid("run root must be an existing non-symlink directory")
    }
    guard outputPath.path.hasPrefix("/"),
      outputPath.deletingLastPathComponent().resolvingSymlinksInPath() == root,
      !FileManager.default.fileExists(atPath: outputPath.path), !isSymlink(outputPath)
    else {
      throw VerificationError.invalid(
        "output must be a new regular file directly under the run root")
    }
    let authorization = try loadObject(authorizationPath, label: "authorization", root: root)
    let receipt = try loadObject(receiptPath, label: "coordinator receipt", root: root)
    let descriptor = try loadObject(descriptorPath, label: "resource descriptor", root: root)
    let health = try loadObject(healthReportPath, label: "health report", root: root)
    let authErrors = Authorization.validateAuthorization(authorization, context: context)
    guard authErrors.isEmpty else {
      throw VerificationError.invalid(
        "authorization is invalid: \(authErrors.joined(separator: "; "))")
    }
    let grants = (authorization["action_grants"] as? [[String: Any]] ?? []).filter {
      $0["grant_id"] as? String == grantID
    }
    guard grants.count == 1 else {
      throw VerificationError.invalid("grant_id must resolve exactly once")
    }
    let grant = grants[0]
    if let exactTarget = grant["target"] as? String, target != exactTarget {
      throw VerificationError.invalid("target differs from the exact grant target")
    }
    guard Set(receipt.keys) == Authorization.coordinatorReceiptFields else {
      throw VerificationError.invalid("coordinator receipt shape is invalid")
    }
    guard receipt["owner_run_id"] as? String == authorization["run_id"] as? String else {
      throw VerificationError.invalid("coordinator receipt belongs to another run")
    }
    guard receipt["owner_actor"] as? String == authorization["selected_writer"] as? String else {
      throw VerificationError.invalid("coordinator receipt belongs to another writer")
    }
    guard receipt["resource_key"] as? String == grant["resource_key"] as? String else {
      throw VerificationError.invalid("coordinator receipt resource differs from the grant")
    }
    guard !descriptor.isEmpty else {
      throw VerificationError.invalid("resource descriptor must be a non-empty object")
    }
    let normalizedDescriptor = try ResourceCoordinator.normalizeDescriptor(
      resource: receipt["resource"] as? String ?? "", descriptor: descriptor)
    let descriptorDigest = try ResourceCoordinator.descriptorSHA256(
      resource: receipt["resource"] as? String ?? "", descriptor: normalizedDescriptor)
    guard receipt["descriptor_sha256"] as? String == descriptorDigest else {
      throw VerificationError.invalid("resource descriptor digest differs from the receipt")
    }
    guard !paths.isEmpty, Set(paths).count == paths.count, paths.allSatisfy({ !$0.isEmpty }) else {
      throw VerificationError.invalid("paths must be non-empty unique strings")
    }
    let checkpointDigest = try specCheckpointPath.map {
      try Authorization.canonicalSHA256(loadAny($0, label: "Spec Kit checkpoint", root: root))
    }
    let observationDigest = try appleObservationPath.map {
      try Authorization.canonicalSHA256(loadAny($0, label: "Apple observation", root: root))
    }
    let appleAction = try appleActionPath.map {
      try loadObject($0, label: "Apple action", root: root)
    }
    let isApple = (grant["action"] as? String)?.hasPrefix("apple.") == true
    if isApple && appleAction == nil {
      throw VerificationError.invalid("Apple grants require --apple-action")
    }
    if !isApple && appleAction != nil {
      throw VerificationError.invalid("non-Apple grants cannot include --apple-action")
    }
    let spec = authorization["spec_kit"] as? [String: Any]
    let request: [String: Any] = [
      "run_id": authorization["run_id"]!, "authorization_id": authorization["authorization_id"]!,
      "authorization_hash": Authorization.authorizationHash(authorization),
      "delivery_target": authorization["delivery_target"]!,
      "system": grant["system"]!, "action": grant["action"]!, "target": target,
      "grant_id": grant["grant_id"]!,
      "idempotency_key": grant["idempotency_key"]!, "repository": authorization["repository"]!,
      "spec_snapshot_sha256": spec?["snapshot_sha256"] ?? NSNull(), "paths": paths,
      "apple": appleAction ?? NSNull(), "lease_id": receipt["lease_id"]!,
      "lease_owner": receipt["owner_actor"]!,
      "lease_resource": receipt["resource"]!, "lease_resource_key": receipt["resource_key"]!,
      "resource_descriptor": normalizedDescriptor, "coordinator_receipt": receipt,
      "operation": grant["operation"]!, "operation_input": grant["operation_input"]!,
      "constraint_sha256": grant["constraint_sha256"]!,
      "phase": grant["phase"]!, "spec_checkpoint_sha256": checkpointDigest ?? NSNull(),
      "apple_observation_sha256": observationDigest ?? NSNull(),
      "writer_actor": authorization["selected_writer"]!,
      "health_report_sha256": "sha256:" + (try Authorization.canonicalSHA256(health)),
    ]
    guard Set(request.keys) == Authorization.requestFields else {
      throw VerificationError.invalid(
        "generated request field set drifted from the installed contract")
    }
    try createExclusiveJSON(request, output: outputPath)
    return request
  }

  private static func loadAny(_ path: URL, label: String, root: URL) throws -> Any {
    do { return try Authorization.loadStablePrivateJSON(path, root: root) } catch {
      throw VerificationError.invalid("\(label) is not a stable private JSON file: \(error)")
    }
  }
  private static func loadObject(_ path: URL, label: String, root: URL) throws -> [String: Any] {
    guard let result = try loadAny(path, label: label, root: root) as? [String: Any] else {
      throw VerificationError.invalid("\(label) must contain an object")
    }
    return result
  }
  private static func createExclusiveJSON(_ value: Any, output: URL) throws {
    let data =
      try JSONSerialization.data(
        withJSONObject: value, options: [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes])
      + Data("\n".utf8)
    let descriptor = open(output.path, O_WRONLY | O_CREAT | O_EXCL | O_NOFOLLOW, S_IRUSR | S_IWUSR)
    guard descriptor >= 0 else {
      throw VerificationError.invalid(
        "output must be a new regular file directly under the run root")
    }
    var succeeded = false
    defer {
      close(descriptor)
      if !succeeded { unlink(output.path) }
    }
    try data.withUnsafeBytes { bytes in
      var pointer = bytes.baseAddress!
      var remaining = bytes.count
      while remaining > 0 {
        let written = Darwin.write(descriptor, pointer, remaining)
        guard written > 0 else { throw VerificationError.invalid("cannot write action request") }
        pointer += written
        remaining -= written
      }
    }
    guard fsync(descriptor) == 0 else {
      throw VerificationError.invalid("cannot sync action request")
    }
    succeeded = true
  }
  private static func isSymlink(_ url: URL) -> Bool {
    var info = stat()
    return lstat(url.path, &info) == 0 && (info.st_mode & S_IFMT) == S_IFLNK
  }
  private static func isDirectory(_ url: URL) -> Bool {
    var directory: ObjCBool = false
    return FileManager.default.fileExists(atPath: url.path, isDirectory: &directory)
      && directory.boolValue
  }
}

public enum PrepareActionRequestCommand {
  public static func run(arguments: [String], context: RuntimeContext) throws -> Int32 {
    let options = try CLIOptions(arguments)
    guard let authorization = options.url("authorization"), let receipt = options.url("receipt"),
      let descriptor = options.url("resource-descriptor"),
      let health = options.url("health-report"), let output = options.url("output"),
      let runRoot = options.url("run-root"),
      let grantID = options.value("grant-id"), let target = options.value("target")
    else { throw VerificationError.invalid("missing required prepare-action-request argument") }
    do {
      let request = try PrepareActionRequest.prepare(
        authorizationPath: authorization, receiptPath: receipt, descriptorPath: descriptor,
        healthReportPath: health, outputPath: output, runRoot: runRoot, grantID: grantID,
        target: target, paths: options.all("path"),
        specCheckpointPath: options.url("spec-checkpoint"),
        appleActionPath: options.url("apple-action"),
        appleObservationPath: options.url("apple-observation"), context: context)
      printJSON([
        "status": "ok", "output": output.resolvingSymlinksInPath().path,
        "grant_id": request["grant_id"]!, "authorization_hash": request["authorization_hash"]!,
      ])
      return 0
    } catch {
      printJSON(["status": "blocked", "reason": String(describing: error)])
      return 2
    }
  }
}
